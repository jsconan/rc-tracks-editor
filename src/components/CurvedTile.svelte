<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import CurvedBarrier from '../elements/CurvedBarrier.svelte';
    import CurvedElement from '../elements/CurvedElement.svelte';
    import CurvedTileModel from '../models/CurvedTileModel.js';
    import ControlPoints from './ControlPoints.svelte';

    export let barrierChunks;
    export let barrierWidth;
    export let laneWidth;
    export let ratio = 1;
    export let direction = CurvedTileModel.DIRECTION_RIGHT;
    export let angle = 0;
    export let x = 0;
    export let y = 0;
    export let filter;
    export let id = void 0;

    const model = new CurvedTileModel(laneWidth, barrierWidth, barrierChunks, ratio);
    const innerRadius = model.getInnerRadius();
    const outerRadius = model.getOuterRadius();
    const innerChunks = model.getInnerBarrierChunks();
    const outerChunks = model.getOuterBarrierChunks();
    const tileAngle = model.getDirectionAngle(direction);
    const curveAngle = model.getCurveAngle();
    const curveCenter = model.getCurveCenter(x, y);
    const center = model.getCenterCoord(x, y);

    const innerBarrierRadius = innerRadius;
    const innerBarrierX = curveCenter.x + innerBarrierRadius;
    const innerBarrierY = curveCenter.y;
    const outerBarrierRadius = outerRadius - barrierWidth;
    const outerBarrierX = curveCenter.x + outerBarrierRadius;
    const outerBarrierY = curveCenter.y;
</script>

<g class="tile curved-tile" transform="rotate({angle} {x} {y}) rotate({tileAngle} {center.x} {center.y})" {filter} {id}>
    <CurvedElement
        class="ground"
        cx={curveCenter.x}
        cy={curveCenter.y}
        width={model.width}
        radius={innerRadius}
        angle={curveAngle}
        start={0}
    />
    <CurvedBarrier
        chunks={innerChunks}
        width={barrierWidth}
        radius={innerBarrierRadius}
        angle={curveAngle}
        left={innerBarrierX}
        top={innerBarrierY}
        shift={0}
    />
    <CurvedBarrier
        chunks={outerChunks}
        width={barrierWidth}
        radius={outerBarrierRadius}
        angle={curveAngle}
        left={outerBarrierX}
        top={outerBarrierY}
        shift={1}
    />
</g>

<ControlPoints {model} {direction} {angle} {x} {y} />
