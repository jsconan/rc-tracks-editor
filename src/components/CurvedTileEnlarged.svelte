<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import CurvedBarrier from '../elements/CurvedBarrier.svelte';
    import StraightBarrier from '../elements/StraightBarrier.svelte';
    import CurvedTileEnlargedModel from '../models/CurvedTileEnlargedModel.js';
    import ControlPoints from './ControlPoints.svelte';

    export let barrierChunks;
    export let barrierWidth;
    export let laneWidth;
    export let ratio = 1;
    export let direction = CurvedTileEnlargedModel.DIRECTION_RIGHT;
    export let angle = 0;
    export let x = 0;
    export let y = 0;
    export let filter;
    export let id = void 0;

    const model = new CurvedTileEnlargedModel(laneWidth, barrierWidth, barrierChunks, ratio);
    const barrierLength = model.length / barrierChunks;
    const side = model.getCurveSide();
    const innerRadius = model.getInnerRadius();
    const outerRadius = model.getOuterRadius();
    const sideChunks = model.getSideBarrierChunks();
    const innerChunks = model.getInnerBarrierChunks();
    const outerChunks = model.getOuterBarrierChunks();
    const tileAngle = model.getDirectionAngle(direction);
    const curveAngle = model.getCurveAngle();
    const curveCenter = model.getCurveCenter(x, y);
    const center = model.getCenterCoord(x, y);

    const innerCurveStartX = curveCenter.x + innerRadius;
    const innerCurveStartY = curveCenter.y;
    const innerCurveEndX = innerCurveStartX - innerRadius;
    const innerCurveEndY = innerCurveStartY + innerRadius;
    const leftSideEndX = innerCurveEndX;
    const leftSideEndY = innerCurveEndY + model.width;
    const outerCurveStartX = leftSideEndX + side;
    const outerCurveStartY = leftSideEndY;
    const outerCurveEndX = outerCurveStartX + outerRadius;
    const outerCurveEndY = outerCurveStartY - outerRadius;
    const rightSideEndX = outerCurveEndX;
    const rightSideEndY = outerCurveEndY - side;

    const innerBarrierRadius = innerRadius;
    const innerBarrierX = innerCurveStartX;
    const innerBarrierY = innerCurveStartY;
    const outerBarrierRadius = outerRadius - barrierWidth;
    const outerBarrierX = outerCurveEndX - barrierWidth;
    const outerBarrierY = outerCurveEndY;
    const horizontalBarrierX = leftSideEndX;
    const horizontalBarrierY = leftSideEndY - barrierWidth;
    const verticalBarrierX = rightSideEndX - barrierWidth;
    const verticalBarrierY = rightSideEndY;
</script>

<g
    class="tile curved-tile-enlarged"
    transform="rotate({angle} {x} {y}) rotate({tileAngle} {center.x} {center.y})"
    {filter}
    {id}
>
    <path
        class="ground"
        d="M {innerCurveStartX} {innerCurveStartY}
           A {innerRadius} {innerRadius} 0 0 1 {innerCurveEndX} {innerCurveEndY}
           L {leftSideEndX} {leftSideEndY}
           L {outerCurveStartX} {outerCurveStartY}
           A {outerRadius} {outerRadius} 0 0 0 {outerCurveEndX} {outerCurveEndY}
           L {rightSideEndX} {rightSideEndY}"
    />
    <CurvedBarrier
        chunks={innerChunks}
        width={barrierWidth}
        radius={innerBarrierRadius}
        angle={curveAngle}
        left={innerBarrierX}
        top={innerBarrierY}
        shift={0}
    />
    <CurvedBarrier
        chunks={outerChunks}
        width={barrierWidth}
        radius={outerBarrierRadius}
        angle={curveAngle}
        left={outerBarrierX}
        top={outerBarrierY}
        shift={1}
    />
    <StraightBarrier
        chunks={sideChunks}
        width={barrierWidth}
        length={barrierLength}
        left={horizontalBarrierX}
        top={horizontalBarrierY}
        shift={0}
        vertical={false}
    />
    <StraightBarrier
        chunks={sideChunks}
        width={barrierWidth}
        length={barrierLength}
        left={verticalBarrierX}
        top={verticalBarrierY}
        shift={1}
        vertical={true}
    />
</g>

<ControlPoints {model} {direction} {angle} {x} {y} />
