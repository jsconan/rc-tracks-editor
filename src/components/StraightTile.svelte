<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import StraightBarrier from '../elements/StraightBarrier.svelte';
    import StraightElement from '../elements/StraightElement.svelte';
    import { TILE_DIRECTION_RIGHT } from '../helpers/types';

    export let model;
    export let direction = TILE_DIRECTION_RIGHT;
    export let angle = 0;
    export let x = 0;
    export let y = 0;
    export let filter = void 0;
    export let id = void 0;

    const width = model.getWidth();
    const height = model.getLength();
    const chunks = model.getSideBarrierChunks();
    const innerRadius = model.getInnerRadius();
    const outerRadius = model.getOuterRadius();
    const tileAngle = model.getDirectionAngle(direction);
    const curveCenter = model.getCurveCenter(x, y);
    const center = model.getCenterCoord(x, y);
    const barrierLength = model.length / model.barrierChunks;
    const barrierWidth = model.barrierWidth;
    const vertical = true;

    const leftBarrierX = curveCenter.x + innerRadius;
    const leftBarrierY = curveCenter.y;
    const rightBarrierX = curveCenter.x + outerRadius - barrierWidth;
    const rightBarrierY = curveCenter.y;

    $: rotation = angle ? `rotate(${angle} ${x} ${y})` : '';
    $: orientation = tileAngle ? `rotate(${tileAngle} ${center.x} ${center.y})` : '';
    $: transform = rotation || orientation ? `${rotation}${orientation}` : '';
</script>

<g class="tile straight-tile" {transform} {filter} {id}>
    <StraightElement class="ground" x={leftBarrierX} y={leftBarrierY} {width} {height} />
    <StraightBarrier
        {chunks}
        width={barrierWidth}
        length={barrierLength}
        left={leftBarrierX}
        top={leftBarrierY}
        shift={0}
        {vertical}
    />
    <StraightBarrier
        {chunks}
        width={barrierWidth}
        length={barrierLength}
        left={rightBarrierX}
        top={rightBarrierY}
        shift={1}
        {vertical}
    />
</g>
