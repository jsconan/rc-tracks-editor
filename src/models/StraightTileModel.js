/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { STRAIGHT_TILE_TYPE } from '../helpers/types.js';
import TileModel from './TileModel.js';

/**
 * Represents a straight track tile.
 */
export default class StraightTileModel extends TileModel {
    /**
     * Computes the actual width of the tile with respect to the ratio.
     * @returns {number}
     */
    getWidth() {
        return this.width;
    }

    /**
     * Computes the angle for rotating the tile to the left.
     * @returns {number}
     */
    getDirectionAngleLeft() {
        return 180;
    }
}

/**
 * The type of tile.
 * @constant {string} StraightTileModel.TYPE
 */
Object.defineProperty(StraightTileModel, 'TYPE', {
    value: STRAIGHT_TILE_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
