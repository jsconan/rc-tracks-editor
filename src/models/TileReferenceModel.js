/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import getTileComponent from '../helpers/getTileComponent.js';
import getTileModel from '../helpers/getTileModel.js';
import { isDirectionValid, isTypeValid, STRAIGHT_TILE_TYPE, TILE_DIRECTION_RIGHT } from '../helpers/types.js';
import uid from '../helpers/uid.js';

/**
 * Represents a reference to a tile model.
 */
export class TileReferenceModel {
    /**
     * Represents a reference to a tile model with the given specifications.
     * @param {string} type - The type of referenced tile.
     * @param {string} direction - The direction of the tile, it can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     */
    constructor(type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        this.id = uid();
        this.setType(type);
        this.setDirection(direction);
        this.setRatio(ratio);
    }

    /**
     * The identifier for the model.
     * @type {string}
     */
    get modelId() {
        return `${this.type}-${this.ratio}`;
    }

    /**
     * Sets the type of referenced tile.
     * @param {string} type - The type of referenced tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given type is not valid.
     */
    setType(type) {
        if (!isTypeValid(type)) {
            throw new TypeError('A valid type of tile is needed!');
        }

        this.type = type;
        return this;
    }

    /**
     * Sets the direction of the tile.
     * It can be either TileModel.DIRECTION_RIGHT or TileModel.DIRECTION_LEFT.
     * @param {string} direction - The direction of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given direction is not valid.
     */
    setDirection(direction) {
        if (!isDirectionValid(direction)) {
            throw new TypeError('A valid direction is needed!');
        }

        this.direction = direction;
        return this;
    }

    /**
     * Sets the size factor of the tile.
     * 1 means the tile fits 1 tile section in each direction.
     * @param {number} ratio - The size factor of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     */
    setRatio(ratio) {
        this.ratio = Math.abs(ratio || 1);
        return this;
    }

    /**
     * Creates the model of the tile with the given size constraints with respect to the stored reference.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @returns {TileModel} - Returns a tile model of the expected type.
     * @throws {TypeError} - If the given specifications object is not valid.
     */
    getModel(specs) {
        const Model = getTileModel(this.type);
        // @ts-expect-error
        return new Model(specs, this.direction, this.ratio);
    }

    /**
     * Gets the component constructor of the tile with respect to the stored reference.
     * @returns {SvelteComponent} - Returns the constructor of the component.
     */
    getComponent() {
        return getTileComponent(this.type);
    }

    /**
     * Computes the coordinates of the tile.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {tileCoord} - The computed coordinates for rendering the tile.
     * @throws {TypeError} - If the given specifications object is not valid.
     */
    build(specs, x = 0, y = 0, angle = 0) {
        const { id, type } = this;
        const component = this.getComponent();
        const model = this.getModel(specs);
        const outputAngle = model.getOutputAngle(angle);
        const outputCoord = model.getOutputCoord(x, y, angle);
        const centerCoord = model.getCenterCoord(x, y, angle);

        return { id, type, x, y, angle, centerCoord, outputCoord, outputAngle, model, component };
    }
}

/**
 * @typedef {object} tileCoord - Represents a tile ready to be rendered.
 * @property {string} id - The identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {Vector2D} centerCoord - The coordinates of the center of the tile.
 * @property {Vector2D} outputCoord - The coordinates of the output of the tile.
 * @property {number} outputAngle - The rotation angle at the output of the tile.
 * @property {TileModel} model - The tile model with respect to its type.
 * @property {SvelteComponent} component - The constructor of the component.
 */

/**
 * @typedef {import('./TileSpecifications').TileSpecifications} TileSpecifications
 */

/**
 * @typedef {import('./TileModel').TileModel} TileModel
 */

/**
 * @typedef {import('./Vector2D').Vector2D} Vector2D
 */

/**
 * @typedef {import('svelte').SvelteComponent} SvelteComponent
 */
