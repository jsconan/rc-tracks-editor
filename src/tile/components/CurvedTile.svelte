<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { getContext } from 'svelte';
    import { tileParameters, groundColor } from '../helpers';
    import { TileSpecifications } from '../config';
    import { CurvedBarrier, CurvedElement } from '../elements';
    import { CurvedTileModel } from '../models';

    export let direction = CurvedTileModel.DIRECTION_RIGHT;
    export let ratio = 1;
    export let angle = 0;
    export let x = 0;
    export let y = 0;
    export let filter = void 0;
    export let id = void 0;

    const specs = getContext(TileSpecifications.CONTEXT_ID);
    const barrierWidth = specs.barrierWidth;
    const width = specs.width;

    $: model = new CurvedTileModel(specs, direction, ratio);
    $: tile = tileParameters.curve(model, x, y);
    $: transform = model.getRotateTransform(x, y, angle);
</script>

<g class="tile curved-tile" {transform} {filter} {id}>
    <CurvedElement
        cx={tile.curveCenter.x}
        cy={tile.curveCenter.y}
        {width}
        radius={tile.innerRadius}
        angle={tile.curveAngle}
        start={0}
        {...groundColor()}
    />
    <CurvedBarrier
        chunks={tile.innerChunks}
        width={barrierWidth}
        radius={tile.innerRadius}
        angle={tile.curveAngle}
        left={tile.innerX}
        top={tile.innerY}
        shift={0}
    />
    <CurvedBarrier
        chunks={tile.outerChunks}
        width={barrierWidth}
        radius={tile.outerRadius}
        angle={tile.curveAngle}
        left={tile.outerX}
        top={tile.outerY}
        shift={1}
    />
</g>
