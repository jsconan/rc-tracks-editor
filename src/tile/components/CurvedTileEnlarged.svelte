<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { getContext } from 'svelte';
    import { groundColor } from '../helpers';
    import { TileSpecifications } from '../config';
    import { CurvedElementEnlarged, CurvedBarrier, StraightBarrier } from '../elements';
    import { CurvedTileEnlargedModel } from '../models';

    export let direction = CurvedTileEnlargedModel.DIRECTION_RIGHT;
    export let ratio = 1;
    export let angle = 0;
    export let x = 0;
    export let y = 0;
    export let filter = void 0;
    export let id = void 0;

    const specs = getContext(TileSpecifications.CONTEXT_ID);
    const barrierLength = specs.barrierLength;
    const barrierWidth = specs.barrierWidth;
    const width = specs.width;

    /**
     * Computes the parameters for rendering the tile at the expected position.
     * @param {CurvedTileEnlargedModel} model
     * @param {number} tileX
     * @param {number} tileY
     * @returns {object}
     * @private
     */
    function getTileParameters(model, tileX, tileY) {
        const side = model.getCurveSide();
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius() - barrierWidth;
        const sideChunks = model.getSideBarrierChunks();
        const innerChunks = model.getInnerBarrierChunks();
        const outerChunks = model.getOuterBarrierChunks();
        const curveAngle = model.getCurveAngle();
        const curveCenter = model.getCurveCenter(tileX, tileY);
        const outerBarrier = width - barrierWidth;

        const { x: innerX, y: innerY } = curveCenter.addScalarX(innerRadius);
        const { x: outerX, y: outerY } = curveCenter.addCoord(
            side + outerRadius,
            innerRadius + outerBarrier - outerRadius
        );
        const { x: horizontalX, y: horizontalY } = curveCenter.addScalarY(innerRadius + outerBarrier);
        const { x: verticalX, y: verticalY } = curveCenter.addScalarX(innerRadius + outerBarrier);

        return {
            side,
            innerRadius,
            outerRadius,
            curveAngle,
            curveCenter,
            sideChunks,
            innerChunks,
            outerChunks,
            innerX,
            innerY,
            outerX,
            outerY,
            horizontalX,
            horizontalY,
            verticalX,
            verticalY
        };
    }

    $: model = new CurvedTileEnlargedModel(specs, direction, ratio);
    $: tile = getTileParameters(model, x, y);
    $: transform = model.getRotateTransform(x, y, angle);
</script>

<g class="tile curved-tile-enlarged" {transform} {filter} {id}>
    <CurvedElementEnlarged
        cx={tile.curveCenter.x}
        cy={tile.curveCenter.y}
        {width}
        side={tile.side}
        radius={tile.innerRadius}
        {...groundColor()}
    />
    <CurvedBarrier
        chunks={tile.innerChunks}
        width={barrierWidth}
        radius={tile.innerRadius}
        angle={tile.curveAngle}
        left={tile.innerX}
        top={tile.innerY}
        shift={0}
    />
    <CurvedBarrier
        chunks={tile.outerChunks}
        width={barrierWidth}
        radius={tile.outerRadius}
        angle={tile.curveAngle}
        left={tile.outerX}
        top={tile.outerY}
        shift={1}
    />
    <StraightBarrier
        chunks={tile.sideChunks}
        width={barrierWidth}
        length={barrierLength}
        left={tile.horizontalX}
        top={tile.horizontalY}
        shift={0}
        vertical={false}
    />
    <StraightBarrier
        chunks={tile.sideChunks}
        width={barrierWidth}
        length={barrierLength}
        left={tile.verticalX}
        top={tile.verticalY}
        shift={1}
        vertical={true}
    />
</g>
