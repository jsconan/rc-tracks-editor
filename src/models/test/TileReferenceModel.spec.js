/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_LEFT,
    TILE_DIRECTION_RIGHT
} from '../../helpers/types';
import CurvedTile from '../../tiles/CurvedTile.svelte';
import CurvedTileEnlarged from '../../tiles/CurvedTileEnlarged.svelte';
import StraightTile from '../../tiles/StraightTile.svelte';
import CurvedTileModel from '../CurvedTileModel.js';
import CurvedTileEnlargedModel from '../CurvedTileEnlargedModel.js';
import StraightTileModel from '../StraightTileModel.js';
import TileReferenceModel from '../TileReferenceModel.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;

describe('TileReferenceModel', () => {
    it('is a class', () => {
        expect(TileReferenceModel).toEqual(expect.any(Function));
    });

    describe('can build a reference to a tile', () => {
        it('with the default specifications', () => {
            const ref = new TileReferenceModel();

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
        });

        it('with the given type', () => {
            const ref = new TileReferenceModel(STRAIGHT_TILE_TYPE);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
        });

        it('with the given direction', () => {
            const ref = new TileReferenceModel(STRAIGHT_TILE_TYPE, TILE_DIRECTION_LEFT);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
        });

        it('with the given ratio', () => {
            const ref = new TileReferenceModel(STRAIGHT_TILE_TYPE, TILE_DIRECTION_LEFT, 2);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
        });

        it('but throws error', () => {
            expect(() => new TileReferenceModel('')).toThrow('A valid type of tile is needed!');
            expect(() => new TileReferenceModel(STRAIGHT_TILE_TYPE, -1)).toThrow('A valid direction is needed!');
        });
    });

    describe('can set', () => {
        it('the type of tile', () => {
            const ref = new TileReferenceModel();

            expect(ref.type).toBe(STRAIGHT_TILE_TYPE);
            expect(ref.setType(CURVED_TILE_TYPE)).toBe(ref);
            expect(ref.type).toBe(CURVED_TILE_TYPE);
            expect(() => ref.setType('')).toThrow('A valid type of tile is needed!');
        });

        it('the direction of the tile', () => {
            const ref = new TileReferenceModel();

            expect(ref.direction).toBe(TILE_DIRECTION_RIGHT);
            expect(ref.setDirection(TILE_DIRECTION_LEFT)).toBe(ref);
            expect(ref.direction).toBe(TILE_DIRECTION_LEFT);
            expect(() => ref.setDirection(-1)).toThrow('A valid direction is needed!');
        });

        it('the ratio of the tile', () => {
            const ref = new TileReferenceModel();

            expect(ref.ratio).toBe(1);
            expect(ref.setRatio(2)).toBe(ref);
            expect(ref.ratio).toBe(2);
            expect(ref.setRatio(-1).ratio).toBe(1);
            expect(ref.setRatio(0).ratio).toBe(1);
            expect(ref.setRatio(0.5).ratio).toBe(0.5);
        });
    });

    describe('can build the model for', () => {
        it('a straight tile', () => {
            const ref = new TileReferenceModel(STRAIGHT_TILE_TYPE);
            const tile = ref.getModel(laneWidth, barrierWidth, barrierChunks);

            expect(tile).toBeInstanceOf(StraightTileModel);
            expect(tile).toMatchSnapshot();
        });

        it('a curved tile', () => {
            const ref = new TileReferenceModel(CURVED_TILE_TYPE);
            const tile = ref.getModel(laneWidth, barrierWidth, barrierChunks);

            expect(tile).toBeInstanceOf(CurvedTileModel);
            expect(tile).toMatchSnapshot();
        });

        it('a curved tile enlarged', () => {
            const ref = new TileReferenceModel(CURVED_TILE_ENLARGED_TYPE);
            const tile = ref.getModel(laneWidth, barrierWidth, barrierChunks);

            expect(tile).toBeInstanceOf(CurvedTileEnlargedModel);
            expect(tile).toMatchSnapshot();
        });
    });

    describe('can get the component for', () => {
        it('a straight tile', () => {
            const ref = new TileReferenceModel(STRAIGHT_TILE_TYPE);

            expect(ref.getComponent()).toBe(StraightTile);
        });

        it('a curved tile', () => {
            const ref = new TileReferenceModel(CURVED_TILE_TYPE);

            expect(ref.getComponent()).toBe(CurvedTile);
        });

        it('a curved tile enlarged', () => {
            const ref = new TileReferenceModel(CURVED_TILE_ENLARGED_TYPE);

            expect(ref.getComponent()).toBe(CurvedTileEnlarged);
        });
    });
});
