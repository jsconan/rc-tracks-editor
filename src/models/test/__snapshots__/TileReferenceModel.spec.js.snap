// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileReferenceModel can build a reference to a tile with the default specifications 1`] = `
TileReferenceModel {
  "direction": "right",
  "id": "id-0",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given direction 1`] = `
TileReferenceModel {
  "direction": "left",
  "id": "id-2",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given ratio 1`] = `
TileReferenceModel {
  "direction": "left",
  "id": "id-3",
  "ratio": 2,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given type 1`] = `
TileReferenceModel {
  "direction": "right",
  "id": "id-1",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build the model for a curved tile 1`] = `
CurvedTileModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can build the model for a curved tile enlarged 1`] = `
CurvedTileEnlargedModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can build the model for a straight tile 1`] = `
StraightTileModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "component": [Function],
  "id": "id-15",
  "model": CurvedTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 135,
  "outputCoord": Vector2D {
    "x": 22.21825406947977,
    "y": 100,
  },
  "type": "curved-tile",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "component": [Function],
  "id": "id-14",
  "model": CurvedTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 90,
  "outputCoord": Vector2D {
    "x": -55,
    "y": 55,
  },
  "type": "curved-tile",
  "x": 0,
  "y": 0,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile enlarged at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "component": [Function],
  "id": "id-17",
  "model": CurvedTileEnlargedModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 135,
  "outputCoord": Vector2D {
    "x": 22.21825406947977,
    "y": 100,
  },
  "type": "curved-tile-enlarged",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile enlarged at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "component": [Function],
  "id": "id-16",
  "model": CurvedTileEnlargedModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 90,
  "outputCoord": Vector2D {
    "x": -55,
    "y": 55,
  },
  "type": "curved-tile-enlarged",
  "x": 0,
  "y": 0,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a straight tile at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "component": [Function],
  "id": "id-13",
  "model": StraightTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 45,
  "outputCoord": Vector2D {
    "x": 22.218254069479784,
    "y": 177.78174593052023,
  },
  "type": "straight-tile",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a straight tile at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "component": [Function],
  "id": "id-12",
  "model": StraightTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 0,
  "outputCoord": Vector2D {
    "x": 0,
    "y": 110,
  },
  "type": "straight-tile",
  "x": 0,
  "y": 0,
}
`;
